File: C/Sort/InsertionSort.c
/*===================================================================
==================    METADATA OF THE FILE      =====================
===================================================================*/
/**
 * @author  Rosas Hernandez Oscar Andres
 * @author  Alan Enrique Ontiveros Salazar
 * @author  Laura Andrea Morales 
 * @version 0.1
 * @team    CompilandoConocimiento
 * @date    4/03/2018
 */


/*===================================================================
==================      INSERTION SORT          =====================
===================================================================*/

/**
 * This is a really intuitive sorting algorithm, to do so we say
 * that we will create a new subarray.
 * So the subarray [0, 1] is already sorted, now:
 * Take the next element contiguos to the subarray an put it where it
 * belongs and move all the other 1 place.
 * Now my new subarray is from [0, 2], and repeat, you get the point
 * 
 * @param Data      A pointer to the array of int to sort
 * @param DataSize  The size of the Data array
 * @return          Nothing...I'm modifying the raw data
 */
void InsertionSort(int Data[], int DataSize){                       //= INSERTION SORT ===

	for (int i = 1; i < DataSize; ++i) {                            //Traverse each item

		int j = i;                                                  //A[0..i-1] is already sorted
		int Temp = Data[i];                                         //Let's find next item

		while (j > 0 && Temp < Data[j - 1]) {                       //Move elements of the subarray
			Data[j] = Data[j - 1];                                  //to one element ahead
			j--;                                                    //until we find the correct place
		}

		Data[j] = Temp;                                             //Put there the Temp
	}
}

File: Python/Sort/InsertionSort.py
#====================================================================
#==================      INSERTION SORT          ====================
#====================================================================
def InsertionSort(Data):                                            #= SELECTION SORT ===
    """
    This is a really intuitive sorting algorithm, to do so we say
    that we will create a new subarray.
    So the subarray [0, 1] is already sorted, now:
    Take the next element contiguos to the subarray an put it where it
    belongs and move all the other 1 place.
    Now my new subarray is from [0, 2], and repeat, you get the point
    
    Args:
        Data (list of int / doubles):  A list to Sort

    Returns:
        List of int / doubles: The sorted list
    """

    SortedData = list(Data)

    for i in range(1, len(Data)):                                   #Traverse each item

        Key = SortedData[i];                                        #Lets put the key
        j = i - 1;                                                  #Let m = A[0..i-1]

        while (j >= 0 and Key < SortedData[j]):                     #Move elements of m
            SortedData[j + 1] = SortedData[j];                      #To one element ahead
            j = j - 1;                                              #ForEach element of m

        SortedData[j + 1] = Key;                                    #Put there Datasize

    return SortedData                                               #Return the sorted data



# =========================================
# ===========   TEST SCRIPT     ===========
# =========================================

A = []

print("Give me all the intergers you want, finish in -1")

while (True):
    x = int(input())
    if x == -1: break
    A.append(x)

B = InsertionSort(A)

print(F"Original Data is {A}")
print(F"Sorted Array is {B}")